### PR TITLE
Fix arithmetic overflows

### PR DESCRIPTION
**Pull request by lyakh**
Fix data range checks, account for possible arithmetic overflows

### CODE REVIEW

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -1033,22 +1033,18 @@ static int sof_control_load_volume(struct snd_soc_component *scomp,
...
-		goto out;
-	}
+	if (le32_to_cpu(mc->num_channels) > SND_SOC_TPLG_MAX_CHAN)
+		return -EINVAL;
 
 	/* init the volume get/put data */
 	scontrol->size = struct_size(scontrol->control_data, chanv,
 				     le32_to_cpu(mc->num_channels));
 	scontrol->control_data = kzalloc(scontrol->size, GFP_KERNEL);
-	if (!scontrol->control_data) {
```
Comment (dbaluta) [inline sound/soc/sof/topology.c]: Out of curiosity and laziness to check the code who will free scontrol->control_data?

Comment (lyakh) [reply sound/soc/sof/topology.c]: @dbaluta `sof_control_unload()` I suppose

File: sound/soc/sof/control.c [hunk 2]
```diff
@@ -229,14 +229,16 @@ int snd_sof_bytes_get(struct snd_kcontrol *kcontrol,
 		return -EINVAL;
```
Comment (kv2019i) [inline sound/soc/sof/control.c:229]: typo in commit: Fis multiple -> Fix multiple

**Review by kv2019i — APPROVED** [review]
Thanks @lyakh ! One typo in commit msg, but actual code is good and especially the overflow checks are very welcome.

File: sound/soc/sof/topology.c (hunk 1 above)
Comment (dbaluta) [reply sound/soc/sof/topology.c]: @lyakh yup.

File: sound/soc/sof/control.c (hunk 2 above)
Comment (lyakh) [reply sound/soc/sof/control.c:229]: @kv2019i thanks, fixed

**Comment by plbossart** [comment]
@lyakh I don't like this change, because it keeps quite a few

````
                ret = -EINVAL;
		goto out_free;
````

So for consistency it's better to keep the goto instead of mixing return and gotos.

File: sound/soc/sof/control.c [hunk 3]
```diff
@@ -229,14 +229,16 @@ int snd_sof_bytes_get(struct snd_kcontrol *kcontrol,
...
 
-	size = data->size + sizeof(*data);
-	if (size > be->max) {
+	/* be->max has been verified to be >= sizeof(struct sof_abi_hdr) */
+	if (data->size > be->max - sizeof(*data)) {
 		dev_err_ratelimited(scomp->dev,
-				    "error: DSP sent %zu bytes max is %d\n",
-				    size, be->max);
+				    "error: DSP claims to send %u bytes of data max is %lu\n",
+				    data->size, be->max - sizeof(*data));
```
Comment (ranj063) [inline sound/soc/sof/control.c:236]: @lyakh this message is not quite right. There is actually no DSP involved here at all. We're only passing the cached value in the control

**Comment by lyakh** [comment]
> @lyakh I don't like this change, because it keeps quite a few
> 
> ```
>                 ret = -EINVAL;
> 		goto out_free;
> ```
> 
> So for consistency it's better to keep the goto instead of mixing return and gotos.

@plbossart this is the accepted practice in the kernel - use `return` for first several error exits, where no clean up is needed, and use `goto` for cases where resource freeing is needed. Just checking some random examples:
- audio core:
  - `snd_pcm_dev_register()`, `_snd_pcm_new()` in sound/core/pcm.c
  - `snd_register_device()` in sound/core/sound.c
- ASoC core:
  - `soc_new_pcm_runtime()` in sound/soc/soc-core.c
- driver core:
  - `device_link_add()`, `device_add_attrs()` in drivers/base/core.c
  - `really_probe()` in drivers/base/dd.c

Really, I just opened random files and searched for `goto` and in *every single* example so far I saw `return` used directly where no cleanup was needed and `goto` only when clean up was needed. Not a *single* case of a label immediately before `return`. I'm not saying there aren't any such cases in the kernel, but in my selection this morning I haven't come across a *single* one, and I think that's a rather good selection. Can we comply, please?

**Comment by plbossart** [comment]
> > @lyakh I don't like this change, because it keeps quite a few
> > ```
> >                 ret = -EINVAL;
> > 		goto out_free;
> > ```
> > 
> > 
> > So for consistency it's better to keep the goto instead of mixing return and gotos.
> 
> @plbossart this is the accepted practice in the kernel - use `return` for first several error exits, where no clean up is needed, and use `goto` for cases where resource freeing is needed. Just checking some random examples:
> 
> * audio core:
>   
>   * `snd_pcm_dev_register()`, `_snd_pcm_new()` in sound/core/pcm.c
>   * `snd_register_device()` in sound/core/sound.c
> * ASoC core:
>   
>   * `soc_new_pcm_runtime()` in sound/soc/soc-core.c
> * driver core:
>   
>   * `device_link_add()`, `device_add_attrs()` in drivers/base/core.c
>   * `really_probe()` in drivers/base/dd.c
> 
> Really, I just opened random files and searched for `goto` and in _every single_ example so far I saw `return` used directly where no cleanup was needed and `goto` only when clean up was needed. Not a _single_ case of a label immediately before `return`. I'm not saying there aren't any such cases in the kernel, but in my selection this morning I haven't come across a _single_ one, and I think that's a rather good selection. Can we comply, please?

Historically we've had tons of issues when missing returns and gotos. You are right that a mix is accepted, but I still find it confusing.

I will go with the majority of reviewers on this one.

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -1145,19 +1145,22 @@ static int sof_control_load_bytes(struct snd_soc_component *scomp,
 	struct snd_soc_tplg_bytes_control *control =
 		container_of(hdr, struct snd_soc_tplg_bytes_control, hdr);
 	struct soc_bytes_ext *sbe = (struct soc_bytes_ext *)kc->private_value;
-	int max_size = sbe->max;
+	size_t max_size = sbe->max, priv_size = le32_to_cpu(control->priv.size);
```
Comment (juimonen) [inline sound/soc/sof/topology.c]: would put this comma separated definition/assignment to separate line

Comment (plbossart) [reply sound/soc/sof/topology.c]: +1, I don't think we use commas anywhere

**Pull request merged**